src/repository.rs

```diff
@@ -1,23 +1,58 @@
 //! Repositories are abstraction over a specific mongo collection for a given `Model`
 
 use crate::{CollectionConfig, Model};
-use mongodb::bson::{from_bson, to_bson, Bson, Document};
+use mongodb::bson::oid::ObjectId;
+use mongodb::bson::{doc, from_bson, to_bson, Bson, Document};
 use mongodb::options::*;
 use mongodb::results::*;
+use serde::Deserialize;
+use std::borrow::Borrow;
+
+#[derive(Debug)]
+pub struct BulkUpdate {
+    pub query: Document,
+    pub update: Document,
+    pub options: Option<UpdateOptions>,
+}
+
+impl AsRef<BulkUpdate> for BulkUpdate {
+    fn as_ref(&self) -> &BulkUpdate {
+        &self
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct BulkUpdateResult {
+    #[serde(rename = "n")]
+    pub nb_affected: u64,
+    #[serde(rename = "nModified")]
+    pub nb_modified: u64,
+    #[serde(default)]
+    pub upserted: Vec<BulkUpdateUpsertResult>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct BulkUpdateUpsertResult {
+    pub index: u64,
+    #[serde(alias = "_id")]
+    pub id: ObjectId,
+}
 
 /// Associate a `mongodb::Collection` and a specific `Model`.
 ///
 /// This type can safely be copied and passed around because `std::sync::Arc` is used internally.
 /// Underlying `mongodb::Collection` can be retrieved at anytime with `Repository::get_underlying`.
 #[derive(Debug)]
 pub struct Repository<M: Model> {
+    db: mongodb::Database,
```

Comment:
Small reminder for our future self

```suggestion
    db: mongodb::Database, // FIXME: temporary keep reference to database object for `bulk_update` operation
```